<launch>
  <arg name="use_rviz" default="false" />
  <arg name="use_stereo_odometry" default="true" />
  <arg name="use_rtabmap" default="true" />
  <arg name="odom_topic" default="/odom" /> <!-- only used when use_stereo_odometry is false -->
  
  <!-- frame_id settings -->
  <arg name="base_frame_id" default="BODY" />
  <arg name="odom_frame_id" default="odom_init" />
  <arg name="map_frame_id" default="map" />
  <arg name="publish_tf" default="true" />

  <!-- image topic settings -->
  <arg name="left_image" default="/multisense_local/left/image_rect"/>
  <arg name="right_image" default="/multisense_local/right/image_rect"/>
  <arg name="left_camera_info" default="/multisense_local/left/camera_info"/>
  <arg name="right_camera_info" default="/multisense_local/right/camera_info"/>

  <!-- <param name="use_sim_time" type="bool" value="True"/> -->

  <!-- Just to uncompress images for stereo_image_rect -->
  <!-- <node name="republish_left"  type="republish" pkg="image_transport" -->
  <!--       args="compressed in:=/stereo_camera/left/image_raw_throttle raw out:=/stereo_camera/left/image_raw_throttle_relay" /> -->
  <!-- <node name="republish_right" type="republish" pkg="image_transport" -->
  <!--       args="compressed in:=/stereo_camera/right/image_raw_throttle raw out:=/stereo_camera/right/image_raw_throttle_relay" /> -->

  <!-- Run the ROS package stereo_image_proc for image rectification -->
  <!-- <group ns="/stereo_camera" > -->
  <!--    <node pkg="nodelet" type="nodelet" name="stereo_nodelet"  args="manager"/> -->
  <!--    <node pkg="stereo_image_proc" type="stereo_image_proc" name="stereo_image_proc"> -->
  <!--       <remap from="left/image_raw"    to="left/image_raw_throttle_relay"/> -->
  <!--       <remap from="left/camera_info"  to="left/camera_info_throttle"/> -->
  <!--       <remap from="right/image_raw"   to="right/image_raw_throttle_relay"/> -->
  <!--       <remap from="right/camera_info" to="right/camera_info_throttle"/> -->
  <!--       <param name="disparity_range" value="128"/> -->
  <!--    </node> -->
  <!-- </group> -->

  <group ns="rtabmap">
    <!-- Stereo Odometry -->
    <node pkg="rtabmap_ros" type="stereo_odometry" name="stereo_odometry"
          output="screen" if="$(arg use_stereo_odometry)">
      <remap from="left/image_rect"       to="$(arg left_image)"/>
      <remap from="right/image_rect"      to="$(arg right_image)"/>
      <remap from="left/camera_info"      to="$(arg left_camera_info)"/>
      <remap from="right/camera_info"     to="$(arg right_camera_info)"/>
      <remap from="odom"                  to="stereo_odometry"/>

      <param name="frame_id"      type="string" value="$(arg base_frame_id)"/>
      <param name="odom_frame_id" type="string" value="$(arg odom_frame_id)"/> <!-- broadcasted by /footcoords -->
      <param name="publish_tf"    type="bool"   value="$(arg publish_tf)"/>
      <!-- <param name="initial_pose"  type="string"   value="x y z roll pitch yaw"/> -->
      <!-- <param name="queue_size"    type="int"    value="10"/> -->

      <param name="Odom/Strategy"      type="string" value="0"/> <!-- 0=Frame-to-Map, 1=Frame=to=Frame -->
      <param name="Vis/EstimationType" type="string" value="0"/> <!-- 0=3D->3D 1=3D->2D (PnP) -->
      <param name="Vis/MaxDepth"       type="string" value="10"/>
      <param name="Vis/MinInliers"     type="string" value="10"/>
      <param name="Odom/FillInfoData"  type="string" value="false"/>
      <param name="GFTT/MinDistance"   type="string" value="10"/>
      <param name="GFTT/QualityLevel"  type="string" value="0.00001"/>
    </node>

    <!-- Visual SLAM: args: "delete_db_on_start" and "udebug" -->
    <node name="rtabmap" pkg="rtabmap_ros" type="rtabmap" output="screen"
          args="--delete_db_on_start" if="$(arg use_rtabmap)" respawn="true">
      <param name="frame_id"         type="string" value="$(arg base_frame_id)"/>
      <param name="map_frame_id"     type="string" value="$(arg map_frame_id)"/>
      <param name="odom_frame_id"    type="string" value="$(arg odom_frame_id)"/>
      <param name="publish_tf"       type="bool"   value="$(arg publish_tf)"/>
      <param name="subscribe_stereo" type="bool" value="true"/>
      <param name="subscribe_depth"  type="bool" value="false"/>
      
      <remap from="left/image_rect"       to="$(arg left_image)"/>
      <remap from="right/image_rect"      to="$(arg right_image)"/>
      <remap from="left/camera_info"      to="$(arg left_camera_info)"/>
      <remap from="right/camera_info"     to="$(arg right_camera_info)"/>

      <remap from="odom" to="stereo_odometry" if="$(arg use_stereo_odometry)"/>
      <remap from="odom" to="$(arg odom_topic)" unless="$(arg use_stereo_odometry)"/>
      <remap from="proj_map" to="/map"/>
      
      <param name="queue_size" type="int" value="30"/>
      <!-- <param name="grid_cell_size" value="0.025"/> -->
      <!-- <param name="scan_voxel_size" value="0.025"/> -->
      <!-- <param name="cloud_voxel_size" value="0.025"/> -->

      <!-- RTAB-Map's parameters -->
      <param name="Rtabmap/TimeThr"                   type="string" value="700"/>
      <param name="Kp/MaxFeatures"                    type="string" value="200"/>
      <param name="Kp/MaxDepth"                       type="string" value="10"/>
      <param name="Kp/DetectorStrategy"               type="string" value="0"/>   <!-- use SURF -->
      <param name="SURF/HessianThreshold"             type="string" value="1000"/>
      <param name="Vis/EstimationType"                type="string" value="0"/>   <!-- 0=3D->3D, 1=3D->2D (PnP) -->
      <param name="RGBD/LoopClosureReextractFeatures" type="string" value="true"/>
      <param name="Vis/MaxDepth"                      type="string" value="10"/>
    </node>
  </group>

  <!-- Visualisation RVIZ -->
  <node pkg="rviz" type="rviz" name="rviz"
        args="-d $(find rtabmap_ros)/launch/config/demo_stereo_outdoor.rviz"
        if="$(arg use_rviz)" />

</launch>
